### PR TITLE
Dithering Node

### PR DESCRIPTION
**Pull request by adodge**
Applies a Bayer ordered dithering filter.  (Maybe a little off-mission for an upscaling-oriented project, but a neat and maybe useful effect.)

image
ford_dithered2

### CODE REVIEW

File: backend/src/nodes/nodes/image_adjustment/dither.py
Comment (RunDevelopment) [inline backend/src/nodes/nodes/image_adjustment/dither.py:24]: That function wasn't meant to be exported by the module. Sorry about that.

And this input type isn't meant to be a number type. It's meant to be a distinct type that is incompatible with any other type. But this distinct type has to be declared in `chainner-scope.ts`. Frankly, I wouldn't recommend doing it that way. I would recommend using an `EnumInput` + `option_labels` ([example](https://github.com/chaiNNer-org/chaiNNer/blob/4f4caaca5f3f49769e53620617eb92fff91d0e90/backend/src/nodes/nodes/image_dimension/resize_to_side.py#L106)) because `EnumInput` automatically handles typing for you.

**Comment by theflyingzamboni** [comment]
Dithering is actually something I've been wanting to add at some point, so I don't consider it off-mission. I've seen multiple people who want to redither images after upscaling them.

To that end, would you be willing to expand the scope of this node? Ideally, I'd like a dithering node to have Bayer, Floyd-Steinberg, and Riemersma options (as can be found in ImageMagick), but even just FS would be a good addition. 

Nodes allow for inputs to be displayed/hidden depending on the value of a dropdown input, so if you have a Method dropdown, you could display only the inputs needed for each type, if they differ.

**Comment by joeyballentine** [comment]
This makes me think we should also have a way of extracting color palettes, so you'd be able to dither with specific colors? That might be out of the scope of this PR though

File: backend/src/nodes/impl/dithering.py
Comment (joeyballentine) [inline backend/src/nodes/impl/dithering.py:150]: This should probably be put into a common helper file that can be imported in both places, rather than being copy/pasted

File: backend/src/nodes/nodes/image_adjustment/dither.py
Comment (RunDevelopment) [inline backend/src/nodes/nodes/image_adjustment/dither.py:32]: Oither?

File: backend/src/nodes/impl/dithering.py
Comment (RunDevelopment) [inline backend/src/nodes/impl/dithering.py:423]: That's a lot of code, and it seems like more will come. Maybe we could make one file per dithering method?

File: backend/src/nodes/nodes/image_adjustment/dither.py
Comment (adodge) [reply backend/src/nodes/nodes/image_adjustment/dither.py:32]: Oops.  Wow what an improbable typo.

File: backend/src/nodes/impl/dithering.py
Comment (adodge) [reply backend/src/nodes/impl/dithering.py:423]: Yes, good idea.  I'll break it up

**Comment by adodge** [comment]
Okay, I think this is ready for review.  It supports paletted and uniform quantization with error diffusion (e.g. Floyd-Steinberg) and Riemersma algorithms, and uniform quantization with ordered dithering (e.g. Bayer).  It also supports quantization without dithering, and [some other kinds of error diffusion](https://tannerhelland.com/2012/12/28/dithering-eleven-algorithms-source-code.html), mostly because they were easy to add once I'd done Floyd-Steinberg.

There's also a node for extracting a color palette from an image, using either k-means or median cut methods.

Here's a chain demoing the different options:

[dither_demo.chn.txt](https://github.com/chaiNNer-org/chaiNNer/files/10529623/dither_demo.chn.txt)

Screenshot from 2023-01-29 03-08-54

ford_dither_demo

(I expect that the methods that don't look good in this example probably have other applications where they do better.)

File: backend/src/nodes/nodes/image_channel/palette_extraction.py
Comment (RunDevelopment) [inline backend/src/nodes/nodes/image_channel/palette_extraction.py:64]: I think we can choose a better output type here. Firstly, we know that the returned palette will have a height of 1px. We can't know the width for the "all" method, but for the others it should the palette size, right?
```suggestion
        self.outputs = [
            ImageOutput(
                image_type=expression.Image(
                    width="""
                        match Input1 {
                            PaletteExtractionMethod::All => int(1..),
                            _ => Input2
                        }
                    """, 
                    height=1, 
                    channels_as="Input0"
                )
            ),
        ]
```

File: backend/src/nodes/nodes/image_filter/dither_palette.py
Comment (RunDevelopment) [inline backend/src/nodes/nodes/image_filter/dither_palette.py:39]: This should probably be called "Palette" instead.

File: backend/src/nodes/impl/dithering/constants.py
Comment (RunDevelopment) [inline backend/src/nodes/impl/dithering/constants.py:166]: You did this to get around black messing up the formatting, right? I think a better solution might be to either turn off formatting (`# fmt: off`) for those lines or to use [some of the tricks in this post](https://stackoverflow.com/questions/58584413/black-formatter-ignore-specific-multi-line-code)

**Review by RunDevelopment — COMMENTED** [review]
Awesome work! Thank you @adodge!

I just have a few minor suggestions.

File: backend/src/nodes/nodes/image_channel/palette_extraction.py
Comment (adodge) [reply backend/src/nodes/nodes/image_channel/palette_extraction.py:64]: Nice!

File: backend/src/nodes/nodes/image_filter/dither_palette.py
Comment (adodge) [reply backend/src/nodes/nodes/image_filter/dither_palette.py:39]: Okay.  I was trying to have some consistency with the "Apply LUT" node, but I guess a LUT and a Palette are actually different things, even if they're both 1 row images.

File: backend/src/nodes/impl/dithering/constants.py
Comment (adodge) [reply backend/src/nodes/impl/dithering/constants.py]: Nice.  Yes, this looks a lot better.

File: backend/src/nodes/nodes/image_filter/dither_palette.py
Comment (RunDevelopment) [reply backend/src/nodes/nodes/image_filter/dither_palette.py]: They are both look up tables, but I think that have slightly different purposes. The LUTs for Apply LUT are essentially gradient maps, so the order of colors matters. But with palettes, the order of colors doesn't matter, right?

Comment (adodge) [reply backend/src/nodes/nodes/image_filter/dither_palette.py]: Yeah, that's a good distinction.

**Comment by RunDevelopment** [comment]
@adodge Could you please fix the CI (our linter don't like some of your code)? We can merge this once the CI passes.

**Comment by adodge** [comment]
@RunDevelopment Should be fixed now!

**Comment by adodge** [comment]
So, there are still formatting complaints, but they're coming from files that were changed in the `main` branch.

**Comment by joeyballentine** [comment]
Yeah i think something got screwed up when i merged that last PR. I'll fix that

File: backend/src/nodes/nodes/image_channel/palette_extraction.py
Comment (RunDevelopment) [inline backend/src/nodes/nodes/image_channel/palette_extraction.py:72]: The palette size fix you recently pushed invalidates this type. If the images has, let's say, 3 distinct colors and the user set a palette size of 4, then your fix will return a 3x1px image, right?

Comment (adodge) [reply backend/src/nodes/nodes/image_channel/palette_extraction.py:72]: It should still be 4x1, since I'm padding it:

https://github.com/chaiNNer-org/chaiNNer/blob/eb457466cc68a7d6777ba9a10b3d976ca452ae54/backend/src/nodes/nodes/image_channel/palette_extraction.py#L97-L99

Comment (adodge) [reply backend/src/nodes/nodes/image_channel/palette_extraction.py:72]: This test suggests it's working as intended:  

extracting 10 colors from a 8x1 image:  image

10x1 output:  image

Comment (adodge) [reply backend/src/nodes/nodes/image_channel/palette_extraction.py:72]: Sorry to sneak that in after it's already reviewed.  I'll open another PR next time.

Comment (RunDevelopment) [reply backend/src/nodes/nodes/image_channel/palette_extraction.py:72]: > It should still be 4x1, since I'm padding it

Ah, I didn't see that part. My bad.

> Sorry to sneak that in after it's already reviewed. I'll open another PR next time.

No, don't worry about it. GH automatically dismisses my review when you push new commits, so I would have reviewed them anyway.

**Comment by adodge** [comment]
Looks like `main` is back in order.  Can we rerun the linter here to make sure?  (It passes when I merge and run it locally.)

**Comment by joeyballentine** [comment]
If you merge main to this branch and push it, it'll re-run. If I just re-run it here it will still fail since it doesn't have the fixes from main

**Comment by RunDevelopment** [comment]
> If you merge main to this branch and push it, it'll re-run.

Not until someone with write access approves of them.

**Comment by adodge** [comment]
Okay...  apparently I was on pylint 2.15.10 not 2.16.1, so I wasn't seeing this failure locally.  Now I am 95% sure it will pass in CI.

File: backend/src/nodes/nodes/image_channel/palette_extraction.py
Comment (joeyballentine) [inline backend/src/nodes/nodes/image_channel/palette_extraction.py:67]: Can you give this output a label of "Palette"?

Comment (adodge) [reply backend/src/nodes/nodes/image_channel/palette_extraction.py:67]: Of course!

**Comment by joeyballentine** [comment]
LGTM but just wanna make sure @RunDevelopment is fine with those ignored types

**Comment by RunDevelopment** [comment]
Yeah, that's fine. Pyright can't handle numpy types for some reason. So all you can do is to `# type: ignore` yourself out of this mess.

**Review by RunDevelopment — APPROVED** [review]

**Pull request merged**
